iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -0,0 +1,214 @@
+}
+
+async fn actually_rsync(work: &Work) -> Result<(), std::io::Error> {
+    tracing::error!("running {}", work.src_file);
+    let output = Command::new("rsync")
+        .arg("-a")
+        .arg("-t")
+        .arg("-X")
+        .arg(&work.src_file)
+        .arg(&work.dest_file)
+        .output();
+
+    let output = output.await?;
+
+    tracing::debug!("done {} {}", work.src_file, output.status);
+
+    Ok(())
+}
+
+async fn do_rsync(work_list: &Vec<Work>) -> Result<Vec<FidError>, ImlAgentError> {
+    let result = Vec::new();
+
+    stream::iter(work_list)
+        .map(|xs| Ok::<_, std::io::Error>(xs))
+        .try_for_each_concurrent(100, |fids| {
+            actually_rsync(fids)
+                .or_else(|e| {
+                    println!("error: {}", e);
+                    future::ok(())
+                })
+                .map_ok(|_| println!("done - actually"))
+        })
+        .await?;
+
+    Ok(result)
+}
+
+async fn archive_fids(
+    llapi: LlapiFid,
+    dest_src: &String,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = Vec::new();
+
+    let mut rsync_list: Vec<Work> = Vec::new();
+
+    for fid in fid_list {
+        let res = llapi.fid2path(&fid.fid);
+
+        if res.is_err() {
+            tracing::error!("llapi: failed on fid {}", &fid.fid);
+            continue;
+        }
+
+        let fid_path = res.ok().unwrap();
+
+        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
+        let dest_file = format!("{}/{}", &dest_src, &fid_path);
+        let mut dest_dir = PathBuf::from(&dest_file);
+
+        dest_dir.pop();
+        if dest_dir.is_dir() == false {
+            fs::create_dir_all(&dest_dir).await?;
+        }
+
+        let md = fs::metadata(&src_file).await?;
+        /* invoking mpifileutils is slow, so only do it for directories
+         * and big files, otherwise just use rsync
+         */
+        if md.is_dir() || (md.len() > 1024 * 1024 * 1024) {
+            let output = Command::new("/usr/lib64/openmpi/bin/mpirun")
+                .arg("--hostfile")
+                .arg("/etc/iml/filesync-hostfile")
+                .arg("--allow-run-as-root")
+                .arg("dsync")
+                .arg("-S")
+                .arg(src_file)
+                .arg(dest_file)
+                .output();
+
+            let output = output.await?;
+            result.push(FidError {
+                fid: fid.fid.clone(),
+                data: fid.data.clone(),
+                errno: output.status.code().unwrap_or(0) as i16,
+            });
+            tracing::debug!(
+                "dsync exited with {} {} {}",
+                output.status,
+                std::str::from_utf8(&output.stdout)?,
+                std::str::from_utf8(&output.stderr)?
+            );
+        } else {
+            rsync_list.push(Work {
+                src_file: src_file.clone(),
+                dest_file: dest_file.clone(),
+            });
```

Comment:
```suggestion
            rsync_list.push(Work {
                src_file,
                dest_file,
            });
```

Cloning not needed here.

Reply:
Done